src/ALCProxy.Communication/IProxyServer.cs

```diff
@@ -0,0 +1,157 @@
+        /// </summary>
+        /// <param name="method">the method that needs to be called</param>
+        /// <param name="streams">The parameters for the given method, converted into a serialized object by the client that now need to be deserialized</param>
+        /// <param name="types">The types of each stream, so the server knows how to decode the streams</param>
+        /// <returns></returns>
+        object CallObject(MethodInfo method, IList<object> streams, IList<Type> types);
+    }
+
+    public abstract class ALCServer<I> : IProxyServer
+    {
+        public object instance;
+        public AssemblyLoadContext currentLoadContext;
+        public ALCServer(Type instanceType, Type[] genericTypes, IList<object> serializedConstParams, IList<Type> constArgTypes)
+        {
+            currentLoadContext = AssemblyLoadContext.GetLoadContext(Assembly.GetExecutingAssembly());
+            if (genericTypes != null)
+            {
+                instanceType = instanceType.MakeGenericType(genericTypes.Select(x => ConvertType(x)).ToArray());
+            }
+            constArgTypes = constArgTypes.Select(x => ConvertType(x)).ToList();
+            var constructorParams = DeserializeParameters(serializedConstParams, constArgTypes);
+            SetInstance(instanceType, constArgTypes.ToArray(), constructorParams);
+        }
+        /// <summary>
+        /// Create the instance of the object we want to proxy
+        /// </summary>
+        /// <param name="instanceType">the type of the object we want</param>
+        /// <param name="constructorTypes">The list of types that the constructor of the object takes in as an argument</param>
+        /// <param name="constructorArgs">The physical objects that are the parameters to the constructor</param>
+        protected void SetInstance(Type instanceType, Type[] constructorTypes, object[] constructorArgs)
+        {
+            var ci = instanceType.GetConstructor(constructorTypes);
+            instance = ci.Invoke(constructorArgs);
+        }
+        /// <summary>
+        /// Takes a Type that's been passed from the user ALC, and loads it into the current ALC for use. 
+        /// TODO: Figure out if we actually need to load here, or we can search the ALC for our loaded dependency which may already be loaded.
+        /// </summary>
+        protected Type ConvertType(Type toConvert)
+        {
+            string assemblyPath = Assembly.GetAssembly(toConvert).CodeBase.Substring(8);
+            if (toConvert.IsPrimitive || assemblyPath.Contains("System.Private.CoreLib")) //Can't load/dont want to load extra types from System.Private.CoreLib
+            {
+                return toConvert;
+            }
+            //TODO can we look for the existence of the loaded assembly here, instead of just doing the load?
+            return currentLoadContext.LoadFromAssemblyPath(assemblyPath).GetType(toConvert.FullName);
```

Comment:
The runtime will not load anything twice into the same ALC - it would return the already loaded assembly anyway.
But in general I don't think this code should be here - see my other comments.